libs/mock-builder/src/storage.rs

```diff
@@ -1,64 +1,87 @@
-use std::{any::Any, cell::RefCell, collections::HashMap, fmt};
+use std::{cell::RefCell, collections::HashMap, fmt};
+
+use super::util::TypeSignature;
 
 /// Identify a call in the call storage
 pub type CallId = u64;
 
-trait Callable {
-	fn as_any(&self) -> &dyn Any;
+struct CallInfo {
+	ptr: u128,
+	type_signature: TypeSignature,
 }
 
 thread_local! {
-	static CALLS: RefCell<HashMap<CallId, Box<dyn Callable>>>
-		= RefCell::new(HashMap::default());
-}
-
-struct CallWrapper<Input, Output>(Box<dyn Fn(Input) -> Output>);
-
-impl<Input: 'static, Output: 'static> Callable for CallWrapper<Input, Output> {
-	fn as_any(&self) -> &dyn Any {
-		self
-	}
+	static CALLS: RefCell<HashMap<CallId, CallInfo>> = RefCell::new(HashMap::default());
 }
 
 #[derive(Debug, PartialEq)]
 pub enum Error {
 	CallNotFound,
-	TypeNotMatch,
+	TypeNotMatch(TypeSignature, TypeSignature),
 }
 
 impl fmt::Display for Error {
 	fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-		write!(
-			f,
-			"{}",
-			match self {
-				Error::CallNotFound => "Trying to call a function that is not registered",
-				Error::TypeNotMatch => "The function is registered but the type mismatches",
-			}
-		)
+		match self {
+			Error::CallNotFound => write!(f, "Trying to call a function that is not registered"),
+			Error::TypeNotMatch(expected, found) => write!(
+				f,
+				"The function is registered but the type mismatches. Expected {}, found: {}",
+				expected, found
+			),
+		}
 	}
 }
 
 /// Register a call into the call storage.
 /// The registered call can be uniquely identified by the returned `CallId`.
-pub fn register_call<F: Fn(Args) -> R + 'static, Args: 'static, R: 'static>(f: F) -> CallId {
+pub fn register_call<F: Fn(I) -> O + 'static, I, O>(f: F) -> CallId {
+	let f = Box::new(f) as Box<dyn Fn(I) -> O>;
+	let ptr = Box::into_raw(f);
+
+	// SAFETY: transforming a wide pointer to an u128 is always safe.
```

Comment:
Generally, I don't think this is true. Can you show a short proof?

Reply:
It's safe in the same way a "normal" reference is always transformable into a `* const u64`. In fact, the following line does not require unsafe rust:

```rust
let reference: &u64 = ...
let ptr = reference as *const u64 
```

This is safe because, it is just an address, a number. What it's not safe is what you do with that address, because all invariants and protections are broken after this.

This is why the other unsafe is the critical one, and you must ensure that the ptr still exists and corresponds to the same type as the original.

Reply:
Regarding "why exactly `u128`?" It's because any `&dyn Type` is a wide pointer: one part corresponds to the pointer itself and the other to the metadata pointer that contains the trait information of the closure. We do not care about how both pointers are sorted in memory, because we are undoing the same we did in the other method for the same type.

Reply:
Agree 👍🏻  I'll add a better explanation. Regarding the 2 `usize` fields, I think separating could confuse in the sense that seems that they should be used separately when not. We only need to allocate that, but not use the internals. Again, I think a good explanation into the code is needed.